zone/parcels.cpp

```diff
@@ -0,0 +1,725 @@
+
+	data->item_slot = parcel.item_slot;
+	strn0cpy(data->send_to, parcel.send_to, sizeof(data->send_to));
+
+	worldserver.SendPacket(out);
+}
+
+void Client::DoParcelRetrieve(ParcelRetrieve_Struct parcel_in)
+{
+	auto merchant = entity_list.GetNPCByID(parcel_in.merchant_entity_id);
+	if (!merchant) {
+		SendParcelRetrieveAck();
+		return;
+	}
+
+	auto p = parcels.find(parcel_in.parcel_slot_id);
+	if (p != parcels.end()) {
+		uint32 item_id       = parcel_in.parcel_item_id;
+		uint32 item_quantity = p->second.quantity;
+		if (!item_id || !item_quantity) {
+			LogError("Attempt to retrieve parcel with erroneous item id or quantity for client character id {}.",
+					 CharacterID());
+			SendParcelRetrieveAck();
+			return;
+		}
+
+		auto inst = database.CreateItem(item_id, item_quantity);
+		if (!inst) {
+			SendParcelRetrieveAck();
+			return;
+		}
+
+		switch (parcel_in.parcel_item_id) {
+			case PARCEL_MONEY_ITEM_ID: {
+				AddMoneyToPP(p->second.quantity, true);
+				MessageString(
+					Chat::Yellow,
+					PARCEL_DELIVERED,
+					merchant->GetCleanName(),
+					inst->DetermineMoneyStringForParcels(p->second.quantity).c_str(),
+					p->second.from_name.c_str()
+				);
+				break;
+			}
+			default: {
+				auto free_id = GetInv().FindFreeSlot(false, false);
+				if (CheckLoreConflict(inst->GetItem())) {
+					if (RuleB(Parcel, DeleteOnDuplicate)) {
+						MessageString(
+							Chat::Yellow,
+							PARCEL_DUPLICATE_DELETE,
+							inst->GetItem()->Name
+						);
+					}
+					else {
+						MessageString(
+							Chat::Yellow,
+							DUP_LORE
+						);
+						SendParcelRetrieveAck();
+						return;
+					}
+				}
+				else if (inst->IsStackable()) {
+					inst->SetCharges(item_quantity);
+					if (TryStacking(inst, ItemPacketTrade, true, false)) {
+						MessageString(
+							Chat::Yellow,
+							PARCEL_DELIVERED_2,
+							merchant->GetCleanName(),
+							std::to_string(item_quantity).c_str(),
+							inst->GetItem()->Name,
+							p->second.from_name.c_str()
+						);
+					}
+					else if (free_id != INVALID_INDEX) {
+						inst->SetCharges(item_quantity);
+						if (PutItemInInventory(free_id, *inst, true)) {
+							MessageString(
+								Chat::Yellow,
+								PARCEL_DELIVERED_2,
+								merchant->GetCleanName(),
+								std::to_string(item_quantity).c_str(),
+								inst->GetItem()->Name,
+								p->second.from_name.c_str()
+							);
+						}
+					}
+					else {
+						MessageString(
+							Chat::Yellow,
+							PARCEL_INV_FULL,
+							merchant->GetCleanName()
+						);
+						SendParcelRetrieveAck();
+						return;
+					}
+				}
+				else if (free_id != INVALID_INDEX) {
+					inst->SetCharges(item_quantity > 0 ? item_quantity : 1);
+					if (PutItemInInventory(free_id, *inst, true)) {
+						MessageString(
+							Chat::Yellow,
+							PARCEL_DELIVERED,
+							merchant->GetCleanName(),
+							inst->GetItem()->Name,
+							p->second.from_name.c_str()
+						);
+					}
+					else {
+						MessageString(
```

Comment:
This is short enough to stay on one line

```cpp
MessageString(Chat::Yellow, PARCEL_INV_FULL, merchant->GetCleanName());
```

Reply:
Resolved